File: src/svc/systemcontroller/client_archive.go
package systemcontroller

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
)

// --- Archive ---

// UploadArchive uploads and extracts an archive into the named subvolume.
// Supported formats are tar.gz, tar.bz2, and tar.xz (detected from the
// filename extension). The archive data is read from archiveReader and sent
// as a multipart form upload. When subpath is non-empty, extraction is
// limited to that directory within the subvolume. When stopService is
// non-empty, the named systemd service is stopped before extraction and
// restarted afterward.
func (c *SystemdClient) UploadArchive(ctx context.Context, subvolume string, archiveReader io.Reader, filename, subpath, stopService string) (_ *ArchiveUploadResponse, err error) {
	pr, pw := io.Pipe()

	writer := multipart.NewWriter(pw)
	go func() {
		if err := writer.WriteField("subvolume", subvolume); err != nil {
			pw.CloseWithError(err)
			return
		}
		if subpath != "" {
			if err := writer.WriteField("subpath", subpath); err != nil {
				pw.CloseWithError(err)
				return
			}
		}
		if stopService != "" {
			if err := writer.WriteField("stop_service", stopService); err != nil {
				pw.CloseWithError(err)
				return
			}
		}
		part, err := writer.CreateFormFile("archive", filename)
		if err != nil {
			pw.CloseWithError(err)
			return
		}
		if _, err := io.Copy(part, archiveReader); err != nil {
			pw.CloseWithError(err)
			return
		}
		pw.CloseWithError(writer.Close())
	}()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.route("storage/upload-archive"), pr)
	if err != nil {
		return nil, fmt.Errorf("%w: POST storage/upload-archive: %w", ErrNewRequest, err)
	}
	req.Header.Set("Content-Type", writer.FormDataContentType())
	if c.Token != "" {
		req.Header.Set("Authorization", "Bearer "+c.Token)
	}

	resp, err := c.HTTP.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: POST storage/upload-archive: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "POST", "storage/upload-archive")
	}

	var result ArchiveUploadResponse
	return &result, json.NewDecoder(resp.Body).Decode(&result)
}

// DownloadArchive creates an archive of the specified paths within the named
// subvolume and returns a reader for the archive data. The format parameter
// selects the compression: "tar.gz", "tar.bz2", or "tar.xz". When stopService
// is non-empty, the named systemd service is stopped during archival. The
// caller must close the returned [io.ReadCloser].
func (c *SystemdClient) DownloadArchive(ctx context.Context, subvolume string, paths []string, stopService, format string) (_ io.ReadCloser, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, DownloadArchiveRequest{Subvolume: subvolume, Paths: paths, StopService: stopService, Format: format})

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.route("storage/download-archive"), pr)
	if err != nil {
		return nil, fmt.Errorf("%w: POST storage/download-archive: %w", ErrNewRequest, err)
	}
	req.Header.Set("Content-Type", "application/json")
	if c.Token != "" {
		req.Header.Set("Authorization", "Bearer "+c.Token)
	}

	resp, err := c.HTTP.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%w: POST storage/download-archive: %w", ErrHTTPRequest, err)
	}

	if resp.StatusCode != http.StatusOK {
		defer func() {
			err = errors.Join(err, resp.Body.Close())
		}()
		return nil, readProblemDetail(resp, "POST", "storage/download-archive")
	}

	return resp.Body, nil
}

File: ui/src/api/client-archive.js
import { ApiError, SystemControllerClient } from './core.js'

/**
 * Upload an archive to a subvolume. Supported formats are tar.gz, tar.bz2,
 * and tar.xz (detected from the filename extension). Uses FormData with
 * fetch() directly (not this.post()) so the browser sets the multipart
 * boundary automatically.
 * @param {string} subvolume - Target subvolume name.
 * @param {File} file - Archive file to upload.
 * @param {string} [subpath] - When provided, extract into this subdirectory within the subvolume.
 * @param {string} [stopService] - When provided, stop this systemd service before extraction and restart afterward.
 * @returns {Promise<{needs_restart: boolean, message: string}>}
 */
SystemControllerClient.prototype.uploadArchive = async function (subvolume, file, subpath, stopService) {
  const form = new FormData()
  form.append('subvolume', subvolume)
  if (subpath) form.append('subpath', subpath)
  if (stopService) form.append('stop_service', stopService)
  form.append('archive', file)

  /** @type {HeadersInit} */
  const headers = {}
  if (this.token) {
    headers['Authorization'] = `Bearer ${this.token}`
  }

  const resp = await fetch(`${this.baseURL}/storage/upload-archive`, {
    method: 'POST',
    headers,
    body: form,
  })
  if (resp.status !== 200) {
    const text = await resp.text().catch(() => '')
    throw new ApiError('POST', '/storage/upload-archive', resp.status, text)
  }
  return resp.json()
}

/**
 * Download an archive of the specified subvolume.
 * Returns the raw Response so the caller can stream resp.body to disk.
 * @param {string} subvolume - Source subvolume name.
 * @param {string[]} [paths] - Paths within the subvolume to include; omit to include everything.
 * @param {string} [stopService] - When provided, stop this systemd service during archival.
 * @param {string} [format] - Compression format: "tar.gz" (default), "tar.bz2", "tar.xz"
 * @returns {Promise<Response>}
 */
SystemControllerClient.prototype.downloadArchive = async function (subvolume, paths, stopService, format) {
  /** @type {HeadersInit} */
  const headers = { 'Content-Type': 'application/json' }
  if (this.token) {
    headers['Authorization'] = `Bearer ${this.token}`
  }

  const body = { subvolume }
  if (paths && paths.length > 0) body.paths = paths
  if (stopService) body.stop_service = stopService
  if (format) body.format = format

  const resp = await fetch(`${this.baseURL}/storage/download-archive`, {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  })
  if (resp.status !== 200) {
    const text = await resp.text().catch(() => '')
    throw new ApiError('POST', '/storage/download-archive', resp.status, text)
  }
  return resp
}
